enterprise/server/certgenerator/certgenerator.go

```diff
@@ -0,0 +1,170 @@
+	sshCertDuration   = flag.Duration("certgenerator.ssh.validity", 12*time.Hour, "How long the generated certificate will be valid.")
+	sshCertPrincipals = flag.String("certgenerator.ssh.principals", "", "Comma separated list of principals to include in the generated certificate.")
+
+	caKeyFile = flag.String("certgenerator.ca_key_file", "", "Path to a PEM encoded certificate authority key file used to issue temporary user certificates.")
+	caKey     = flag.String("certgenerator.ca_key", "", "PEM encoded certificate authority key used to issue temporary user certificates.", flag.Secret)
+)
+
+const (
+	// Allow user to get an interactive prompt when using SSH.
+	allowPty = "permit-pty"
+)
+
+type generator struct {
+	verifier *oidc.IDTokenVerifier
+
+	sshSigner ssh.Signer
+}
+
+func (g *generator) Generate(ctx context.Context, req *cgpb.GenerateRequest) (*cgpb.GenerateResponse, error) {
+	idToken, err := g.verifier.Verify(ctx, req.GetToken())
+	if err != nil {
+		return nil, status.PermissionDeniedErrorf("invalid token: %s", err)
+	}
+	claims := struct {
+		Subject       string `json:"sub"`
+		Domain        string `json:"hd"`
+		Email         string `json:"email"`
+		EmailVerified bool   `json:"email_verified"`
+	}{}
+	if err := idToken.Claims(&claims); err != nil {
+		return nil, status.PermissionDeniedErrorf("could not parse claims: %s", err)
+	}
+	if claims.Domain != *domain {
+		return nil, status.PermissionDeniedErrorf("invalid domain %q", claims.Domain)
+	}
+	if !claims.EmailVerified {
+		return nil, status.PermissionDeniedError("email not verified")
+	}
+
+	pk, _, _, _, err := ssh.ParseAuthorizedKey([]byte(req.GetSshPublicKey()))
+	if err != nil {
+		return nil, status.InvalidArgumentErrorf("could not parse public key: %s", err)
+	}
+
+	log.Infof("Generating certificate for: %+v", claims)
+
+	cert := ssh.Certificate{
+		Key:             pk,
+		CertType:        ssh.UserCert,
+		KeyId:           claims.Email,
+		ValidPrincipals: strings.Split(*sshCertPrincipals, ","),
+		ValidAfter:      uint64(time.Now().Unix()),
+		ValidBefore:     uint64(time.Now().Add(*sshCertDuration).Unix()),
+	}
+	cert.Permissions.Extensions = map[string]string{allowPty: ""}
+	if err := cert.SignCert(rand.Reader, g.sshSigner); err != nil {
+		return nil, err
+	}
+
+	return &cgpb.GenerateResponse{
+		SshCert: string(ssh.MarshalAuthorizedKey(&cert)),
+	}, nil
+}
+
+func newGenerator(ctx context.Context) (*generator, error) {
+	provider, err := oidc.NewProvider(ctx, *tokenIssuer)
+	if err != nil {
+		return nil, status.UnavailableErrorf("could not create OIDC provider: %s", err)
+	}
+	key, err := ssl.LoadCertificateKey(*caKeyFile, *caKey)
+	if err != nil {
+		return nil, status.FailedPreconditionErrorf("could not load CA certificate: %s", err)
+	}
+	s, err := ssh.NewSignerFromKey(key)
+	if err != nil {
+		return nil, err
+	}
+	g := &generator{
+		verifier:  provider.Verifier(&oidc.Config{ClientID: *tokenClientID}),
+		sshSigner: s,
+	}
+	return g, nil
+}
+
+func main() {
```

Comment:
I think it's probably worth adding a basic test with a valid token and invalid token, if it's not too difficult (could maybe use selfauth as the oidc provider?)

Reply:
Didn't seem that straightforward, but I'll take another look in a bit.